### ISSUE TITLE
KMS handshake with Version

### ISSUE THREAD

**Issue by liamsi** [tendermint/tendermint#2665]
@ValarDragon suggested to remove the `Version` field from the Canonical(Vote/Proposal) altogether and let the KMS (instead of the HSM) make sure (via an initial handshake) that it doesn't sign any message while "talking to a different version". 

This would have the following advantages:
- less communication overhead when sending Votes/Proposals
- less bytes to sign inside the HSM 
- no need to change Vote/Proposal structs (or the Signable interface)

This would change the offsets though (cc @jleni) 
(related: https://github.com/tendermint/tendermint/issues/2545)

**Comment by ebuchman** [comment]
Added "Remove Version from CanonicalXxx" to https://github.com/tendermint/tendermint/issues/2468. This issue can track performing a handshake between KMS and Tendermint and passing a version through there

**Comment by Liamsi** [comment]
Can't this be closed? We probably do not need a handshake anymore but rather make sure that the messages passed to the kms have a message type the kms knows of (it should check if they are valid when decoding).

**Comment by Liamsi** [comment]
Actually, from the messages exchanged between tendermint and the kms, only votes contain the `SignMsgType` as a field. The HSM module (or whatever provider behind signatory) which processes / signs the canoncialized version of all messages. Not sure if we should fail earlier if the schema changed or not. An easy solution to this would be to add a `Type` field to proposals (and heartbeats) too. Then, we basically "versioned" everything that is sent between tm <-> kms.

**Comment by ebuchman** [comment]
So we send the KMS the non-canonicalized structs, which contain signer info (address, index), but not the ChainID.

We should align the Vote/Proposal fields with the Canonicalized fields so eg. Proposal will have SignMsgType, but I think we still want a handshake. I don't know how multi tenacy will work with the KMS yet (both across keys and across chains), but I would expect to authorize some kind of session up front for a particular key and a particular chain (and version of the protocol), and then be able to send messages that should continue to include all of that information and be checked each time.

In summary, we should consider

- aligning Vote/Proposal fields with canonical order and fields
- send chain-id to the kms along with vote
- introduce a handshake the sets a validator and a chain-id and verify it with every message before signing

**Comment by Liamsi** [comment]
> send chain-id to the kms along with vote

Would this be another breaking change as this would change the messages that get exchanged between remote signer / kms and tendermint?

**Comment by jleni** [comment]
@Liamsi ping me in case we need to adjust the validator app specs/implementation.
Maybe it is a good idea to link an issue here (https://github.com/tendermint/ledger-validator-app) if there are breaking changes.

**Comment by Liamsi** [comment]
I don't think this would break the validator app specs. Only the messages send between tm <-> kms/any remote signer. What gets signed in the end by the app shouldn't be affected by this.

**Comment by jleni** [comment]
> I don't think this would break the validator app specs. Only the messages send between tm <-> kms/any remote signer. What gets signed in the end by the app shouldn't be affected by this.

I asked because of this:

> This would change the offsets though (cc @jleni) (related: #2545)

**Comment by Liamsi** [comment]
I understand. But the chainId would not be included in the vote directly, but (additionally to the Vote) in the SignVoteRequest, or, only once during handshake (still needs to be fledged out).
